crates/tags/src/signed.rs

```diff
@@ -0,0 +1,242 @@
+// SPDX-FileCopyrightText: 2022 Profian Inc. <opensource@profian.com>
+// SPDX-License-Identifier: Apache-2.0
+
+use std::ops::{Deref, DerefMut};
+use std::str::FromStr;
+
+use drawbridge_http::http::Request;
+use drawbridge_http::{async_trait, FromRequest};
+use drawbridge_jose::jws::Jws;
+use drawbridge_type::Entry;
+
+use http_types;
+use serde::{Deserialize, Serialize};
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub struct TagName(String);
+
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct TagValue {
+    pub body: Vec<u8>,
+    pub kind: Kind,
+    pub name: TagName,
+}
+
+impl TagValue {
+    #[allow(dead_code)]
+    pub const TYPE_ENTRY: &'static str = "application/vnd.drawbridge.tag.v1+json";
+    #[allow(dead_code)]
+    pub const TYPE_JOSE: &'static str = "application/jose+json";
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+// value of the HashMap, name is the Key
+#[serde(untagged)]
+pub enum Kind {
+    Signed(Jws),
+    Unsigned(Entry),
+}
+
+impl Deref for TagName {
+    type Target = String;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl DerefMut for TagName {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl FromStr for TagName {
+    type Err = ();
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(TagName(s.to_string()))
+    }
+}
+
+#[async_trait]
+impl FromRequest for TagName {
+    async fn from_request(req: &mut Request) -> http_types::Result<Self> {
+        let mut url_parts: Vec<&str> = req.url().path().split("/").collect();
+        url_parts.reverse();
+        Ok(TagName(url_parts.get(0).unwrap().clone().to_string()))
```

Comment:
We definitely cannot `unwrap()` here, since that would cause a server-side panic on invalid request - we should error out instead.
We should, however validate the tag contents, i.e.:
- validate it's length, perhaps `[1;255]` is a very generous range we could allow here?
- ensure the tag contents are URL-safe, you can take a look at https://github.com/profianinc/drawbridge/blob/ff86631f4e57714574a4c4bcad87e36e73a5265d/crates/name/src/lib.rs#L27-L57 for an example. For now I guess it's okay to just check https://doc.rust-lang.org/std/string/struct.String.html#method.is_ascii I will refactor this later and add more robust validation

Feel free to file an issue and a TODO here and we can address the validation later